File: python/568.py
"""Project Euler Problem 568: Reciprocal Games II.

D(n) = J_B(n) - J_A(n) = H_n / 2^n, where H_n is the n-th harmonic number.

We need the 7 most significant digits of D(123456789) after removing leading zeros.

Approach: compute log10(D(n)) = log10(H_n) - n*log10(2) using high-precision
arithmetic (mpmath), then extract the significant digits from the fractional part.
"""

from __future__ import annotations

import math
from mpmath import mp, mpf, log10, harmonic


def solve() -> str:
    """Solve Problem 568."""
    mp.dps = 50  # 50 decimal digits of precision
    N = 123456789

    # H_N via Euler-Maclaurin summation (fast for large N in mpmath)
    H_N = harmonic(N)

    # log10(D(N)) = log10(H_N) - N * log10(2)
    log10_H_N = float(log10(H_N))
    N_log10_2 = float(N * log10(mpf(2)))
    log10_D = log10_H_N - N_log10_2

    # Extract 7 most significant digits from the fractional part of log10
    frac = log10_D - math.floor(log10_D)
    significant = 10**frac
    return str(significant).replace(".", "")[:7]


def main() -> str:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/568.c
/*
 * Project Euler Problem 568: Reciprocal Games II
 *
 * D(n) = H_n / 2^n. We need 7 most significant digits of D(123456789).
 * log10(D(n)) = log10(H_n) - n*log10(2).
 * H_n ~ ln(n) + gamma + 1/(2n) - 1/(12n^2) + ... (Euler-Maclaurin)
 * So log10(H_n) = log10(ln(n) + gamma + ...).
 * Then extract the fractional part and get 7 significant digits.
 */
#include <stdio.h>
#include <math.h>

int main(void) {
    long long N = 123456789LL;
    double n = (double)N;

    /* Euler-Mascheroni constant */
    double gamma = 0.5772156649015328606065120900824024310421;

    /* H_n via Euler-Maclaurin: ln(n) + gamma + 1/(2n) - 1/(12n^2) + 1/(120n^4) - ... */
    double inv_n = 1.0 / n;
    double inv_n2 = inv_n * inv_n;
    double H_n = log(n) + gamma
                 + inv_n / 2.0
                 - inv_n2 / 12.0
                 + inv_n2 * inv_n2 / 120.0
                 - inv_n2 * inv_n2 * inv_n2 / 252.0
                 + inv_n2 * inv_n2 * inv_n2 * inv_n2 / 240.0;

    /* log10(D(N)) = log10(H_N) - N * log10(2) */
    double log10_H = log10(H_n);
    double log10_D = log10_H - (double)N * log10(2.0);

    /* Extract fractional part */
    double frac = log10_D - floor(log10_D);
    double significant = pow(10.0, frac);

    /* Get 7 digits */
    long long digits = (long long)(significant * 1000000.0);
    printf("%lld\n", digits);
    return 0;
}
